### PR TITLE
Execution Plan Filtering and UI improvements

### PR DESCRIPTION
**Pull request by aasimkhan30**
Action Bar Seperator:
image

Properties Filter:
image

Properties Filter In Plan Comparison:
image

Top operation Filter:
image

### CODE REVIEW

File: src/sql/base/browser/ui/taskbar/media/actionBar.css [hunk 1]
```diff
@@ -0,0 +1,10 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+.monaco-action-bar.vertical .action-item .action-label.separator {
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/taskbar/media/actionBar.css:6]: Why is this necessary? Have you verified that this isn't going to cause any problems in the other places that use actionbar? You may want to consider scoping this down to just execution plan unless you think this is something that affects all action bars.

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesView.ts [hunk 2]
```diff
@@ -127,6 +152,13 @@ export class ExecutionPlanPropertiesView extends ExecutionPlanPropertiesViewBase
 				props = this.sortPropertiesReverseAlphabetically(props);
 				break;
 		}
+		return props;
+	}
+
+	private convertModelToTableRows(props: azdata.executionPlan.ExecutionPlanGraphElementProperty[], parentIndex: number, rows: { [key: string]: string }[] = []): { [key: string]: string }[] {
+		if (!props) {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesView.ts:159]: If this can be undefined use the proper optional type in the parameter list

**Review by erinstellato-ms — CHANGES_REQUESTED** [review]
Can you change the angle of the Custom Zoom icon to be the same angle (to top left) as the Zoom In and Zoom Out icons?  Then all of the zoom icons go the same direction (and the Find Node will go a different direction).

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesViewBase.ts [hunk 3]
```diff
@@ -99,13 +111,27 @@ export abstract class ExecutionPlanPropertiesViewBase implements IVerticalSashLa
...
-		propertiesContent.appendChild(this._headerActionsContainer);
+		this._searchAndActionBarContainer.appendChild(this._headerActionsContainer);
 		this._headerActions = new ActionBar(this._headerActionsContainer, {
 			orientation: ActionsOrientation.HORIZONTAL, context: this
 		});
 		this._headerActions.pushAction([new SortPropertiesByDisplayOrderAction(), new SortPropertiesAlphabeticallyAction(), new SortPropertiesReverseAlphabeticallyAction()], { icon: true, label: false });
 
+		this._propertiesSearchInputContainer = DOM.$('.table-search');
+		this._propertiesSearchInputContainer.classList.add('codicon', searchIconClassNames);
+		this._propertiesSearchInput = new InputBox(this._propertiesSearchInputContainer, this._contextViewService, {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesViewBase.ts:126]: Register this (and make this class disposable - then make sure it itself is registered by anything that creates it and so on...)

Make sure you go through this class and register any other disposables you create too (like the onDidChange handler below)

**Comment by aasimkhan30** [comment]
> Can you change the angle of the Custom Zoom icon to be the same angle (to top left) as the Zoom In and Zoom Out icons? Then all of the zoom icons go the same direction (and the Find Node will go a different direction).

@erinstellato-ms  updated. 
image

File: src/sql/base/browser/ui/taskbar/media/actionBar.css (hunk 1 above)
Comment (aasimkhan30) [reply src/sql/base/browser/ui/taskbar/media/actionBar.css:6]: Currently, the separator in vertical action bars do not look as intended. This is a targeted fix for that. There isn't any other side effects I have noticed. 

This is how separators looks without this css.

image

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesViewBase.ts [hunk 4]
```diff
@@ -91,21 +102,35 @@ export abstract class ExecutionPlanPropertiesViewBase implements IVerticalSashLa
...
 			orientation: ActionsOrientation.HORIZONTAL, context: this
-		});
+		}));
 		this._headerActions.pushAction([new SortPropertiesByDisplayOrderAction(), new SortPropertiesAlphabeticallyAction(), new SortPropertiesReverseAlphabeticallyAction()], { icon: true, label: false });
 
+		this._propertiesSearchInputContainer = DOM.$('.table-search');
+		this._propertiesSearchInputContainer.classList.add('codicon', searchIconClassNames);
+		this._propertiesSearchInput = this._register(new InputBox(this._propertiesSearchInputContainer, this._contextViewService, {
+			ariaDescription: localize('tableSearchDescription', 'Search properties table'),
+			placeholder: localize('tableSearchPlaceholder', 'Filter for any field...')
```
Comment (lewis-sanchez) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesViewBase.ts:127]: You could also move this to the top of the class as a static property.

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesViewBase.ts [hunk 5]
```diff
@@ -245,6 +272,68 @@ export abstract class ExecutionPlanPropertiesViewBase implements IVerticalSashLa
...
+			this._tableComponent.setData(
+				this.flattenTableData(
+					this.searchNestedTableData(searchString, this._tableData).data,
+					-1)
+			);
+		}
+		this._tableComponent.rerenderGrid();
+	}
+
+	private searchNestedTableData(pattern: string, data: Slick.SlickData[]): { include: boolean, data: Slick.SlickData[] } {
```
Comment (lewis-sanchez) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanPropertiesViewBase.ts:288]: Is pattern a regex for a data value or the target data value? If it's just the target data value than renaming it to something along those lines would make more sense.

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanComparisonPropertiesView.ts [hunk 6]
```diff
@@ -281,11 +281,11 @@ export class ExecutionPlanComparisonPropertiesView extends ExecutionPlanProperti
 				}
 				rows.push(row);
 				if (!isString(topProp.value) && !isString(bottomProp.value)) {
-					this.convertPropertiesToTableRows(topProp.value, bottomProp.value, rows.length - 1, indent + 2, rows);
+					row['treeGridChildren'] = this.convertPropertiesToTableRows(topProp.value, bottomProp.value);
 				} else if (isString(topProp?.value) && !isString(bottomProp.value)) {
-					this.convertPropertiesToTableRows(undefined, bottomProp.value, rows.length - 1, indent + 2, rows);
+					row['treeGridChildren'] = this.convertPropertiesToTableRows(undefined, bottomProp.value);
 				} else if (!isString(topProp.value) && !isString(bottomProp.value)) {
```
Comment (lewis-sanchez) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanComparisonPropertiesView.ts:287]: This `else if` condition is the same as the first `if` statement. Is that intentional?

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanComparisonPropertiesView.ts [hunk 7]
```diff
@@ -281,11 +281,11 @@ export class ExecutionPlanComparisonPropertiesView extends ExecutionPlanProperti
 				}
 				rows.push(row);
 				if (!isString(topProp.value) && !isString(bottomProp.value)) {
-					this.convertPropertiesToTableRows(topProp.value, bottomProp.value, rows.length - 1, indent + 2, rows);
```
Comment (lewis-sanchez) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanComparisonPropertiesView.ts:284]: You can condense all this logic to the following since we only want to invoke convertPropertiestoTableRows when at least 1 or more arguments are not strings:
```
const topPropValue = !isString(topProp.value) ? topProp.value : undefined;
const bottomPropValue = !isString(bottomProp.value) ? bottomProp.value : undefined;
if (!(topPropValue !== undefined && bottomPropValue !== undefined)) {
    this.convertPropertiesToTableRows(topPropValue, bottomPropValue);
}

File: src/sql/workbench/contrib/executionPlan/browser/executionPlanComparisonPropertiesView.ts [hunk 8]
```diff
@@ -281,11 +281,11 @@ export class ExecutionPlanComparisonPropertiesView extends ExecutionPlanProperti
...
+					row['treeGridChildren'] = this.convertPropertiesToTableRows(topProp.value, bottomProp.value);
```
Comment (lewis-sanchez) [inline src/sql/workbench/contrib/executionPlan/browser/executionPlanComparisonPropertiesView.ts:284]: You can condense all this logic to the following since we only want to invoke convertPropertiestoTableRows when at least 1 or more arguments are not strings:

```
const topPropValue = !isString(topProp.value) ? topProp.value : undefined;
const bottomPropValue = !isString(bottomProp.value) ? bottomProp.value : undefined;
if (!(topPropValue !== undefined && bottomPropValue !== undefined)) {
    this.convertPropertiesToTableRows(topPropValue, bottomPropValue);
}

**Review by erinstellato-ms — APPROVED** [review]
Thanks @aasimkhan30, that looks better!  :)

**Review by lewis-sanchez — APPROVED** [review]
